### ISSUE TITLE
Need a way to stop / unregister a citizen

### ISSUE THREAD

**Issue by jcgertig** [Akamaozu/node-supe#8]
Please add a way to stop / unregister a citizen

**Comment by Akamaozu** [comment]
Honestly didn't expect anyone knew about this library, let alone used it enough to be able to make useful suggestions about it.

I use this lib extensively in production, but I haven't found a need to unregister / stop a citizen.

Mind sharing what your usecase is, @jcgertig?

**Comment by jcgertig** [comment]
I want to in one instance use a sub supervisor basically as a thread
control where i can dynamically start instances of the same process to
handle a connection each. My connections are basically proxies to a very
heavy data end point so want a thread per connection. But i need to clean
up when the connection is removed. I found this from your blog post becuase
was looking for erlang supervisor pattern for node.

**Comment by Akamaozu** [comment]
@jcgertig

Let me see if I'm understanding you correctly.

You want to create and control processes  that connect to endpoints that produce a lot of data.

You also want to make sure it correctly cleans up after every request.

Is this correct, @jcgertig?

**Comment by jcgertig** [comment]
Yes and no the data is a direct tcp stream and each connection might be
open for days.

**Comment by Akamaozu** [comment]
Got it.

And you want the supervisor to be able to tell a tcp stream citizen to close, clean up and shut down?

**Comment by jcgertig** [comment]
You have .start I want a .stop that removes the registration on that name
and calls

`citizen.ref.kill('SIGHUP');`

because you store a ref to the spawn child correct.

**Comment by jcgertig** [comment]
also a .unregister method since you have a register so if I have a way to
kill my process I can just unregister it

**Comment by Akamaozu** [comment]
@jcgertig I'm truly sorry about the absence of anything resembling documentation. You could have solved this problem yourself, but you had to create this issue because there were no docs.

Back to your issue.

**There currently isn't anything in supe's api that stops or deregisters a citizen, but building a usable version of that should be straightforward**.

It's really a matter of sending a message / signal from the supervisor to the citizen, and making sure the citizen shuts down when it receives it. There's at least three ways in supe's api to accomplish this.

**Here's how to do it via mail**

```js
// supervisor
var citizen = supervisor.get( 'citizen' );
citizen.mail.send({ action: 'shutdown' });

// citizen
citizen.mail.receive( function( envelope ){
  var mail = envelope.msg,
      from = envelope.from;

  if( mail.action && mail.action === 'shutdown' ) return process.exit();
});
```

One benefit of using supe's mail to send the shutdown signal is that citizens only get one mail at a time. They must acknowledge processing the last sent mail before your shutdown mail is sent.

With mail, you can trivially set it up so the command is processed only it is idle.

**If you don't have this restriction, you might be better off using lighter-weight messaging to get the signal across**.

**Here's an example with notices**.

```js
// supervisor
supervisor.noticeboard.notify( 'shutdown-proxy', proxy_citizen_name );

// citizen
citizen.noticeboard.watch( 'shutdown-proxy', 'shutdown-proxy-' + proxy_citizen_name, function( msg ){
  var proxy_to_shutdown = msg.notice;
  if( citizen_name != proxy_to_shutdown ) return;
  else process.exit();
});   
```

Noticeboard is far less restrictive than mail. It's also broadcast to all citizens supervised by the supervisor whose noticeboard was notified. These properties allow you to create more interesting shutdown systems.

For example, you could have have proxy citizens tag (_owner, source, destination, content-type, expected size etc_) their current stream, and have the supervisor send shutdown notices like this:

```js
supervisor.noticeboard.notify( 'shutdown-proxy', { owner: 'client whose payment attempt just failed' });
```

Now you can trivially shut down every proxy owned by a delinquent client.

**Comment by Akamaozu** [comment]
@jcgertig 

De-registering a citizen is more complicated than shutting it down, so I have a few questions about it.

1. What happens if you call deregister on a currently-running citizen?
2. What happens if there is mail for a citizen you're trying to deregister?
3. What if a citizen is dependent on another citizen and that gets deregistered?

**Comment by jcgertig** [comment]
I had thought your .start and .register should return promises so if a
process just does not even run the first time its not even started because
likely a larger issue like so

```
supervisor.start('a', 'a.js')
.then((a) => { /* do thing with 'a' */ })
.catch((err) => { /* failed immediately */ });

supervisor.register('a', 'a.js')
.then(() => { /* 'a' is registered */ })
.catch((err) => { /* failed immediately like name conflict */ });
```

and the new .stop and .unregister could be the same

```
supervisor.stop('a')
.then(() => { /* a stopped */ })
.catch((err) => { /* error stopping or unregistering 'a' */ });

supervisor.unregister('a')
.then(() => { /* a is unregistered */ })
.catch((err) => { /* failed immediately like 'a' still running */ });
```

**Comment by Akamaozu** [comment]
@jcgertig 

1. In regards to promisified `supervisor.start` and `supervisor.register`:

    - `supervisor.start` and `supervisor.register` are operations that can't fail.

        - `supervisor.start` under the hood uses node core's `child_process.spawn`. There's no callback or return fail instance. It will always give you a child_process. See: https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options

        - `supervisor.register` simply creates an object and adds properties and to it.

    - **Could you explain what value a promise-based api adds here?**

2. In regards to `supervisor.stop` and `supervisor.unregister`: 

    - I'm not particular about api type (promise vs callback). I'd prefer core module apis to be more similar, which right now means callbacks, but **I'm open to a promise-based api version**. 

    - **My concern for creating `supervisor.stop` and `supervisor.deregister` is making the api behavior reliable in all sorts of situations. Here's some I'm concerned about**: https://github.com/Akamaozu/node-supe/issues/8#issuecomment-386805340

**Comment by jcgertig** [comment]
Looking through your code register can fail therefore start can fail. It
can fail if you try to register a name that is already registered. Right
now you just throw errors but with a promise there is a nice way to handle
errors without needing a try catch. If no .catch is added to a handled
promise the errors will still be thrown as unhandled promise errors.

I was saying that deregister should fail if the process has not been
stopped. .stop would force the stop then unregister. As for mail if the
process has been stoped it cant get mail so the mail should just be
flushed. Not sure the api should care about dependencies since there is no
auto deregister so the devlopers should be concerned with handling of
dependencies.

**Comment by Akamaozu** [comment]
Thanks for taking the time to look around supe's internals :)

You're right on a number of things.

**1. Supe methods can/do fail. I'd forgotten about that.**

I let supe errors crash the process since it typically means something is majorly wrong, but ... 

Others (_or me, someday_) might prefer handling supe errors and promises definitely beat try/catch there.

I like the idea of a promise-based supe api. Would you be interested in working on this?

**2. Stop / Deregister logic makes sense**.
I agree with your breakdown except in one little way.

I don't think `supervisor.stop` should automatically deregister after a stop. I feel `supervisor.deregister` should stop before deregistering.

Otherwise, sounds great.

**3. Components shouldn't care about how others clean up**.

Making a `citizen-deregistering` hook means components can hook in and clean up the way they feel is best. 

E.g. Mail can console.log all unprocessed mail about to be dropped.

**Comment by jcgertig** [comment]
Okay I will fork and make a pr is a bit

**Comment by jcgertig** [comment]
@Akamaozu Opened Pr #9

**Comment by jcgertig** [comment]
Also when will you publish the version with deregister?

**Comment by jcgertig** [comment]
@Akamaozu

**Comment by Akamaozu** [comment]
@jcgertig I have no need for `deregister`, so I don't have usecases to test if the API fits in nicely with older methods and if it behaves how you'd expect.

Didn't want to publish it without some feedback from someone who wants it.

Have you tried it?

**Comment by jcgertig** [comment]
As of now its working how I want since pulling from github. If I have issues i will make prs

**Comment by jcgertig** [comment]
I do need this soon if possible

**Comment by jcgertig** [comment]
I have something I want to start using in prod and I dont like linking to git repos for packages. When do you think this can be released?

**Comment by Akamaozu** [comment]
@jcgertig I'll publish it next week :)

**Comment by jcgertig** [comment]
would it be possible for you to release a `0.2.0-rc.1` today? on my work boxes we cant clone git so it has to be in npm

**Comment by jcgertig** [comment]
@Akamaozu ^

**Comment by jcgertig** [comment]
@Akamaozu ?

**Comment by Akamaozu** [comment]
@jcgertig pre-release is a bit complicated for me. published v0.2.0 instead.
